### ISSUE TITLE
Faster FP32 power calculation on CUDA GPUs

### ISSUE THREAD

**Issue by yhtang** [google/jax#12509]
This is related to #8068 but concerns performance rather than accuracy.

@nouiz and I noticed that `jax.numpy.power` currently lowers to nearly 200 PTX instructions. This is similar to how nvcc would compile `powerf` without `-use_fast_math`.

However, the hardware-implemented SP exponential/logarithm functions can compute `a**b` as `exp(b*log(a))` with much better performance and slightly worse accuracy. This corresponds to the `__powf` CUDA C++ intrinsic or `powerf` with `-use_fast_math`.
```
        lg2.approx.ftz.f32      %f3, %f1;
        mul.ftz.f32     %f4, %f2, %f3;
        ex2.approx.ftz.f32      %f5, %f4;
```

Is this something worths further investigation?

**Comment by hawkinsp** [comment]
I think it's a good question. This optimization could certainly be done in XLA and seems worthwhile particularly for neural network users.

The key thing to my mind is we need a better API at both XLA and JAX levels for allowing the user to choose either the "fast" implementation or the "accurate" implementation. If we had that, we would have more flexibility to have both and choose between them. The closest analogue I can think of is the `precision` flag on `Dot` and `Conv` operations in HLO.

**Comment by nouiz** [comment]
@cheshire What do you think about adding an XLA_FLAGS=--use_fast_math that would trigger using the hardware implementation of functions when available?
Like the nvcc --use_fast_math parameter?

At least, it will give a way to try it and see if it give useful speed up or not on different model.

**Comment by yhtang** [comment]
Another option might be to introduce a new [JAX config](https://jax.readthedocs.io/en/latest/jax.config.html) entry, so that users can enable it from within Python (e.g. a notebook) by:
```python
jax.config.update('use_fast_math_cuda', True)
# or alternatively
jax.config.update('use_fast_pow_cuda', True)
```

**Comment by cheshire** [comment]
@nouiz Generally I'm against flags as each flag doubles a possible number of XLA configurations. Then we end up with an exponential number of configurations, only one of them being tested. Fast math flag in particular is not great, as it's not clear what semantics it implies (e.g. losing some precision is generally OK, but incorrect `nan/inf` propagation is unacceptable, fastmath tends to bundle the two)

@hawkinsp for additional thoughts.

In this particular case,

1) Would something like `jax.numpy.fast_power` make sense?
2) How much precision do we lose by converting to `exp(b*log(a))`? Maybe we could even do it by default?

**Comment by yhtang** [comment]
> Would something like `jax.numpy.fast_power` make sense?

Throwing my 2c here: A new ufunc like this need to work on all hardware targets. However, it may not be 'fast' on the CPU, which does not have hardware log/exp instructions. Or maybe an alternative implementation of `power/log/exp` using SSE/AVX could help...

**Comment by rmlarsen** [comment]
Just be aware that when you say "slightly worse accuracy" that computing `pow(x,y) = exp2(y*log2(x))` can lead to several thousand ULPs of error, even when `exp2` and `log2` are accurate to 1 ULP, say. There are arguments where the error magnification is very large. I would definitely recommend an explicit `fast_power` API or equivalent.

**Comment by rmlarsen** [comment]
Another point is special case handling, which is particularly convoluted and ugly to implement for `pow`. See, e.g., the long laundry list of cases to handle here: https://en.cppreference.com/w/cpp/numeric/math/pow

**Comment by hawkinsp** [comment]
Out of curiosity, what was the original workload that calls pow() heavily? Do you have a benchmark?

**Comment by yhtang** [comment]
> Out of curiosity, what was the original workload that calls pow() heavily? Do you have a benchmark?

It is relatively infrequent for deep learning workloads to use `pow()`.

However, kernel methods such as Gaussian process regression may use it heavily because exponentiation with a _learnable exponent_ is a very common trick to tune a kernel. The [rational quadratic kernel](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.RationalQuadratic.html) also uses exponentiation with learnable exponents. To create a kernel matrix of NxN, the `pow()` method will have to be applied elementwise to each of the matrix element so the cost could potentially be significant.

**Comment by hawkinsp** [comment]
I wonder if it might not be cleanest to write this at the Python as `exp(b * log(a))`. That is simple, clear, and it's obvious to the user what to do if they don't like the numerics of the result. XLA should fuse the computations into a single kernel.

**Comment by rmlarsen** [comment]
I'm not sure I like that suggestion. exp(b * log(a)) can be quite
inaccurate, and many people might not be aware of that.

**Comment by hawkinsp** [comment]
To clarify, my suggestion was *the user* writes that. They would need to be aware of it to write it! By default you get an accurate `pow()` if you write `pow()`.

**Comment by rmlarsen** [comment]
I understand. But once the user writes that, we are obliged to give them
the inaccurate result. It's a bad forcing function to push them towards
that.
